.github/workflows/buildpack-ubuntu2004.yml

```diff
@@ -0,0 +1,46 @@
+name: buildpack:ubuntu2004
+
+on:
+  pull_request:
+    paths:
+      - '.github/workflows/buildpack-ubuntu2004.yml'
+      - 'scripts/docker/buildpack/Dockerfile.ubuntu2004'
+jobs:
+  build:
+    env:
+      GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      GITHUB_DOCKER_REPOSITORY: docker.pkg.github.com/${{ github.repository }}
+      IMAGE_NAME: buildpack
+      IMAGE_VARIANT: ubuntu2004
+
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Check version
+        run: .github/workflows/buildpack/check-version.sh
+
+      - name: Build image
+        run: docker build scripts/docker/${IMAGE_NAME} --file scripts/docker/${IMAGE_NAME}/Dockerfile.$IMAGE_VARIANT --tag $IMAGE_NAME
+
+      - name: Test image
+        run: .github/workflows/buildpack/test-image.sh
+
+      - name: Push image
+        run: .github/workflows/buildpack/push-image.sh
```

Comment:
I don't think extracting stuff to scripts like this is really working out. What about just:
```
      - name: Check version
        run: |
          echo "${GITHUB_TOKEN}" | docker login docker.pkg.github.com -u "${GITHUB_ACTOR}" --password-stdin
          .github/workflows/buildpack/update_image.sh ubuntu2004
          docker logout docker.pkg.github.com
```

You can then have debug output via special ``echo``s from the update_image.sh, which gives almost as much feedback as using several steps in the workflows and the update_image.sh script will work quite similarly for all docker files.

Instead of locally tagging the image with specific names, you can just tag them (locally) as "current-image" (or whatever) - then that doesn't need to be in variables either, simplifying things further.

For testing you can just have ``.github/workflows/buildpack/test-ubuntu2004.sh`` scripts and call them as
``test-${IMAGE_KIND}.sh`` and use symlinks when they'd just do the same.

Reply:
I'm not very sure what you mean here. Do you mean to put everything in one single script executed by a single step?

Reply:
Yes, I don't see why not. The docker login and logout is rather specific to the actions, so I'd keep that part there, but everything else is pretty generic and you just need to pass the information which Dockerfile triggered it to one single script.

This will simplify all the actions and also this one script will be way simpler than the separate scripts you have now.

Reply:
To elaborate a bit more: I mean in combination with https://github.com/ethereum/solidity/pull/9368#discussion_r452078991

So you will have one script that takes one argument (e.g. the suffix of the Dockerfile, so ``ubuntu2004`` for ``Dockerfile.ubuntu2004``) and this would do the following:
- check if the diff of ``Dockerfile.$1`` between develop and HEAD increases the version label by one, fail otherwise.
- build a docker image from ``Dockerfile.$1`` and tag it e.g.  "current-image"
- extract the version label from the built image
- run the docker image "current-image" while mounting the repo root to ``/root/project`` and executing the script ``/root/project/.github/workflows/buildpack/test-$1.sh`` in it.
- re-tag and push the docker image to the github packages repository as ``solidity-buildpack-deps:$1-$VERSION``

All of this will be generic and shared across all the Dockerfiles. The only thing you need on top of that are the
``/root/project/.github/workflows/buildpack/test-$1.sh`` scripts. For emscripten this can just be symlinked to ``scripts/travis-emscripten/build_emscripten.sh``. For the ossfuzz image it can be a link to the ``scripts/ossfuzz_build_test.sh`` we want to have that is shared with CircleCI. For the ubuntu images it will probably be identical, so you can have them link to one script like ``scripts/ubuntu_build_test.sh`` that basically does the same as the ``run_build`` part in the CircleCI config. In a second step we should actually make ``run_build`` in CircleCI re-use that very same script as well, but I'd not do that in this PR, but only later in a second step.

Do you think that makes sense? The idea is to keep everything as independent of the actual CI system as possible - and to unify what we use for testing of these images and what is actually run on CircleCI later - and to reduce redundancy between what is done for the separate ``Dockerfile``s, because *basically* we'll do the same for all of them.

Reply:
@ekpyron thx for the clarification! It totally makes sense. I have a first version ready and it's building now. I'm really not a big fan of long running testing processes - it slows down the development speed a lot. However, a drawback for having all steps in one single script is that it makes it very hard to find where exactly something went wrong, if something went wrong. If the test builds passed I will try something to preserve the steps while using the same script.

Reply:
Sounds great!
Regarding finding out exactly what went wrong - the single script can still ``echo`` and print whatever step it is just performing, that alone wouldn't be too bad (yeah, maybe green arrows are nicer, but it's not a must ;-)) - and maybe some of the special github action "workflow command" echo's can create some more structure https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#about-workflow-commands - the core.startGroup/group stuff there looks a bit like it, but I haven't really looked into those or tried them myself yet.

Reply:
@ekpyron I had some troubles building `ubuntu1604.clang.ossfuzz` correctly, but it seem to work now. Probably the solution is still a bit hacky and need some cleanup. However, I found an easy solution to have the different steps back: https://github.com/aarlt/solidity/pull/19. I just read your comment right now, but I agree core.startGroup/core.core.endGroup sounds promising. I would look into this tomorrow.